### PR TITLE
Overhaul overload semantics, remove erasure, add union math

### PR DESCRIPTION
**Pull request by Michael0x2a**
Overhaul overload semantics, remove erasure, add union math

This pull request:

1.  Modifies how mypy handles overloads to match the proposal I made in the typing repo.

2.  Starts removing type erasure from overload checks.

3.  Adds support for basic union math.

4.  Makes overloads respect keyword-only args

This pull request does NOT implement the following:

1.  Special-casing descriptors

2.  Improving how operator methods are handled: I've tweaked one or two methods to take advantage of some of the changes I made in this PR, but stayed away from making any drastic changes for now.

3.  Detecting partially overlapping argument types -- e.g. code like:

        @overload
        def f(x: Union[A, B]) -> str: ...
        @overload
        def f(x: Union[B, C]) -> int: ...

        @overload
        def g(x: Tuple[int, ...]) -> str: ...
        @overload
        def g(x: Tuple[int, int]) -> int: ...

4.  Detecting overlapping "argument counts". For example, we should flag the following as an error since the first alternative could potentially overlap with the second.

        @overload
        def f(*args: int) -> int: ...
        @overload
        def f(x: int, y: int, z: int) -> str: ...

5.  The "is-more-precise" relation. It's working in most normal cases but still contains a few bugs, mostly relating to type vars. For example, this currently isn't being flagged as an error:

        class Wrapper(Generic[T]):
            @overload
            def f(self, x: int) -> int: ...  # No error?
            @overload
            def f(self, x: T) -> str: ...

    (This PR does the right thing if 'T' isn't bound to a containing class though:)

        class Wrapper:
            @overload
            def f(self, x: int, y: int) -> int: ...  # Error
            @overload
            def f(self, x: T, y: T) -> str: ...

    Currently, what I'm doing is using the existing `is_more_precise` method, which calls `is_proper_subtype`. I think i'll either need to rewrite `is_more_precise` to do what I want with typevars or find a way of forcing the two methods to unify their typevars before running the `is_proper_subtype` check.

The plan is to address these 5 TODOs in future pull requests. Items 1 and 2 are basically orthogonal to the overloads overhaul; items 3, 4, and 5 basically boil down to finding ways to teach mypy to detect if one thing is *potentially* compatible with another.

For example, mypy contains code to tell if one type is *definitely* a subtype of another; fixing items 3 and 5 involve writing code to check if a type is *potentially* a subtype of another.

### CODE REVIEW

**Comment by Michael0x2a** [comment]
@ilevkivskyi - I think this is ready to be reviewed whenever you have time. The tests are currently failing, but I that should be fixed once [https://github.com/python/typeshed/pull/2150] lands and we do the whole sync typeshed + rebase dance.

Also, just a note: accepting this PR without implementing todos 3, 4, and 5 will mean mypy will have a hole in the type system, at least temporarily. If the team decides they'd rather avoid that, I don't mind working on those todos first: it's the same amount of work for me, so I guess it comes down to what's easier review/work with.

Another option might be to add an optional flag that throws a warning if a call ends up matching multiple overloads with distinct return types -- it'll give us a mechanism for detecting overloads that end up being unsafe/would be a useful knob for people who prefer a more paranoid typechecker. 

(I also don't think it'll be too noisy: I tried implementing + running that check on mypy and got back only three errors, mostly relating to the zip builtin function.)

**Comment by Michael0x2a** [comment]
So, correction: most of the failing tests are due to typeshed issues that are easy to fix.

However, one of the failures exposed an issue that might be harder to address: it boils down to an unfortunate interaction between non-strict optional + removing type erasure. This came up with Python 2's map function which, when simplified, looks roughly like this:

    @overload
    def map(func: None, 
            iter1: Iterable[T1], 
            iter2: Iterable[T2]) -> Iterable[Tuple[T1, T2]]: ...
            
    @overload
    def map(func: Callable[[T1, T2], S], 
            iter1: Iterable[T1], 
            iter2: Iterable[T2]) -> Iterable[S]: ...

Basically, this is fine with strict optional since the two types for `func` are disjoint. It fails if you disable strict optional though, since None is a subtype of everything.

Previously, mypy didn't catch this since the two return types, when erased, look identical. But now that we're getting rid of type erasure...

Not 100% sure what to do here. I guess we could lie and pretend that strict optional always applies when checking overloads? Or special-case overloads + None somehow? These options seem sort of suboptimal though...

Thoughts?

**Comment by gvanrossum** [comment]
Presumably when someone writes overloads like that they want None to be treated special regardless of the setting of strict_optional. So I think it should be honored. If map(None, ...) is called the overload should use the first alternative. If the argument is not just None, the first alternative should be skipped -- even if the argument type is Optional[Callable[......]], since the Optional is lost when strict_optional is off.

**Comment by ilevkivskyi** [comment]
> I guess we could lie and pretend that strict optional always applies when checking overloads? Or special-case overloads + None somehow?

Either way is fine. The idea if someone writes an overload on `None` one expects it to be a "normal type" (i.e.not magically a subtype of everything). I think we already have some special casing for protocols so that `__hash__ = None` works consistently with Python semantics independently of `--strict-optional`.

**Comment by ilevkivskyi** [comment]
@Michael0x2a I will start reviewing this tomorrow or Monday, in the meantime I created a new label to simplify coordination of work on overloads https://github.com/python/mypy/labels/topic-overloads

File: mypy/checker.py [hunk 1]
```diff
@@ -407,49 +407,63 @@ def _visit_overloaded_func_def(self, defn: OverloadedFuncDef) -> None:
...
         return None
 
     def check_overlapping_overloads(self, defn: OverloadedFuncDef) -> None:
         # At this point we should have set the impl already, and all remaining
         # items are decorators
-        for i, item in enumerate(defn.items):
-            assert isinstance(item, Decorator)
-            sig1 = self.function_type(item.func)
-            for j, item2 in enumerate(defn.items[i + 1:]):
+        with experiments.strict_optional_set(True):
```
[the remark this answers is not in the document]
Comment (Michael0x2a) [reply mypy/checker.py]: Done.

File: mypy/checker.py [hunk 2]
```diff
@@ -407,49 +407,63 @@ def _visit_overloaded_func_def(self, defn: OverloadedFuncDef) -> None:
...
-                impl_type_subst = impl_type
-                if impl_type.variables:
-                    unified = unify_generic_callable(impl_type, sig1, ignore_return=False)
-                    if unified is None:
-                        self.fail("Type variable mismatch between " +
-                                  "overload signature {} and implementation".format(i + 1),
-                                  defn.impl)
+                    if overload_can_never_match(sig1, sig2):
+                        self.msg.overloaded_signature_will_never_match(
+                            i + 1, i + j + 2, item2.func)
```
[the remark this answers is not in the document]
Comment (Michael0x2a) [reply mypy/checker.py]: It's because we're iterating over the remaining overloads using an enumeration: 'j' starts at zero.

(`i + j + 2` was also what the code appeared to be doing before, so I figured it'd be fine to just copy it for the new check.)

File: mypy/messages.py [hunk 3]
```diff
@@ -937,11 +942,19 @@ def incompatible_typevar_value(self,
...
-                                      context: Context) -> None:
+    def overloaded_signatures_overlap(self, index1: int, index2: int, context: Context) -> None:
         self.fail('Overloaded function signatures {} and {} overlap with '
                   'incompatible return types'.format(index1, index2), context)
 
+    def overloaded_signature_will_never_match(self, index1: int, index2: int,
+                                              context: Context) -> None:
+        self.fail(
+            'Overloaded function signature {index2} will never be matched: '
+            'function {index1}\'s parameter type(s) are the same or broader'.format(
```
[the remark this answers is not in the document]
Comment (Michael0x2a) [reply mypy/messages.py]: Done

File: mypy/checker.py [hunk 4]
```diff
@@ -407,49 +407,63 @@ def _visit_overloaded_func_def(self, defn: OverloadedFuncDef) -> None:
...
+                    assert isinstance(impl_type, CallableType)
+                    assert isinstance(sig1, CallableType)
+                    if not is_callable_compatible(impl_type, sig1,
+                                                  is_compat=is_subtype, ignore_return=True):
+                        self.msg.overloaded_signatures_arg_specific(i + 1, defn.impl)
+                    impl_type_subst = impl_type
+                    if impl_type.variables:
+                        unified = unify_generic_callable(impl_type, sig1, ignore_return=False)
+                        if unified is None:
+                            self.fail("Type variable mismatch between " +
```
[the remark this answers is not in the document]
Comment (Michael0x2a) [reply mypy/checker.py]: Yeah, I don't really understand the logic here either. TBH, I didn't look too hard at it: the only changes I made in my first pass was to just indent everything by a level (because of the strict-optional with statement).

I decided to just remove this error message in my second pass: I was able to find only one test case that produced this error message. (And that test case looked ill-defined anyways -- it was checking the `f(x: Any) -> T` case I was asking about on gitter earlier.)

(Same sort of thing for your comment about the three conditions up above: that was old logic I decided to just keep. I did remove the `sig1 is None` bit though -- I think that condition is actually always going to be false.)

File: mypy/checker.py [hunk 5]
```diff
@@ -1832,10 +1846,18 @@ def check_multi_assignment_from_union(self, lvalues: List[Expression], rvalue: E
             # Bind a union of types collected in 'assignments' to every expression.
             if isinstance(expr, StarExpr):
                 expr = expr.expr
-            types, declared_types = zip(*items)
+
+            # TODO: See todo in binder.py, ConditionalTypeBinder.assign_type
+            # It's unclear why the 'declared_type' param is sometimes 'None'
+            clean_items = []  # type: List[Tuple[Type, Type]]
+            for type, declared_type in items:
+                assert declared_type is not None
```
[the remark this answers is not in the document]
Comment (Michael0x2a) [reply mypy/checker.py]: I think what was happening here was that previously, the call to `zip` directly below was incorrectly inferred to have a type of 'Any', which made the call to `self.binder.assign_type` always succeed.

My overload changes made zip start inferring the correct types here, which led to a type error when calling `binder.assign_type`. I thought about adding casts, but was wanted to avoid silently propagating 'None' values into unexpected places as much as possible: I figured a noisy assert would be easier for others to debug if this ever became an issue.

That said, mypy isn't running into any issues here so maybe a cast is safe: let me know and I can make that change.

File: mypy/checker.py [hunk 6]
```diff
@@ -3547,18 +3569,96 @@ def type(self, type: Type) -> Type:
...
-    Two signatures s and t are overlapping if both can be valid for the same
+def is_unsafe_overlapping_overload_signatures(signature: CallableType,
+                                              other: CallableType) -> bool:
+    """Check if two overloaded function signatures may be unsafely overlapping.
+
+    We consider two functions 's' and 't' to be unsafely overlapping both
+    of the following are true:
+
+    1.  s's parameters are all more precise or partially overlapping with t's
+    1.  s's return type is NOT a subtype of t's.
```
[the remark this answers is not in the document]
Comment (Michael0x2a) [reply mypy/checker.py]: Whoops, fixed (same for comment above).

File: mypy/checker.py [hunk 7]
```diff
@@ -3601,12 +3701,11 @@ def is_more_general_arg_prefix(t: FunctionLike, s: FunctionLike) -> bool:
     """Does t have wider arguments than s?"""
     # TODO should an overload with additional items be allowed to be more
     #      general than one with fewer items (or just one item)?
-    # TODO check argument kinds and otherwise make more general
     if isinstance(t, CallableType):
         if isinstance(s, CallableType):
-            t, s = unify_generic_callables(t, s)
```
[the remark this answers is not in the document]
Comment (Michael0x2a) [reply mypy/checker.py]: It's because `is_callable_compatible` will internally call `unify_generic_callables` for us.

File: mypy/checkexpr.py [hunk 8]
```diff
@@ -1104,68 +1106,242 @@ def check_arg(self, caller_type: Type, original_caller_type: Type,
...
+                                                          arg_kinds, arg_names, callable_name,
+                                                          object_type, context, arg_messages)
+        if inferred_result is not None:
+            # Success! Stop early.
+            return inferred_result
+
+        # Step 3: At this point, we know none of the overload alternatives exactly match.
+        #         We fall back to using the erased types to help with union math/help us
+        #         produce a better error message.
+        erased_targets = self.overload_erased_call_targets(plausible_targets, arg_types,
```
[the remark this answers is not in the document]
Comment (Michael0x2a) [reply mypy/checkexpr.py]: Oh, I did it because if an overload has a signature of `(int) -> int` and `(List[int]) -> str`, and if a user tries passing in an argument of type `List[str]` or something, we probably want guess that the user meant the second option and give an error message based off of that.

(This also let us maintain compatibility with a lot of the existing error messages.)

I added a comment to clarify.

File: mypy/checkexpr.py [hunk 9]
```diff
@@ -1104,68 +1106,242 @@ def check_arg(self, caller_type: Type, original_caller_type: Type,
...
+                                   object_type: Optional[Type],
+                                   context: Context,
+                                   arg_messages: Optional[MessageBuilder] = None,
+                                   ) -> Optional[Tuple[Type, Type]]:
+        """Attempts to find the first matching callable from the given list.
+
+        If multiple targets match due to ambiguous Any parameters, returns (AnyType, AnyType).
+        If no targets match, returns None.
+
+        Assumes all of the given targets have argument counts compatible with the caller.
```
[the remark this answers is not in the document]
Comment (Michael0x2a) [reply mypy/checkexpr.py]: Done

File: mypy/checkexpr.py [hunk 10]
```diff
@@ -1104,68 +1106,242 @@ def check_arg(self, caller_type: Type, original_caller_type: Type,
...
+        """
+
+        arg_messages = self.msg if arg_messages is None else arg_messages
+        matches = []  # type: List[CallableType]
+        inferred = []  # type: List[Tuple[Type, Type]]
+        args_contain_any = any(isinstance(arg, AnyType) for arg in arg_types)
```
[the remark this answers is not in the document]
Comment (Michael0x2a) [reply mypy/checkexpr.py]: Oh whoops, I didn't even think of that case. But yeah, I think we should return 'Any' -- I fixed the code + added test cases.

File: mypy/checkexpr.py [hunk 11]
```diff
@@ -1104,68 +1106,242 @@ def check_arg(self, caller_type: Type, original_caller_type: Type,
...
+        expected_names = callables[0].arg_names
+        expected_kinds = callables[0].arg_kinds
+
+        for target in callables:
+            if target.arg_names != expected_names or target.arg_kinds != expected_kinds:
+                # We conservatively end if the overloads do not have the exact same signature.
+                # TODO: Enhance the union overload logic to handle a wider variety of signatures.
+                return None
+
+            if target.is_generic():
```
[the remark this answers is not in the document]
Comment (Michael0x2a) [reply mypy/checkexpr.py]: I ended up just rewriting how we handle generics -- I think I copied the generics code here from somewhere (probably `is_callable_compatible` or something), but found it did a poor job of handling more complicated cases while I was playing around with the "use overloads + NoReturn as type guards" idea.

I also added more comments to the new implementation. I'm still not entirely sure if my new approach is correct, but if it's wrong, it'll hopefully be more obviously wrong :)

File: mypy/checkexpr.py [hunk 12]
```diff
@@ -1104,68 +1106,242 @@ def check_arg(self, caller_type: Type, original_caller_type: Type,
...
+        #     @overload
+        #     def f(x: B, y: A) -> None: ...
+        #
+        # If we continued and synthesize "def f(x: Union[A,B], y: Union[A,B]) -> None: ...",
+        # then we'd incorrectly accept calls like "f(A(), A())" when they really ought to
+        # be rejected.
+        #
+        # However, that means we'll also give up if the original overloads contained
+        # any unions. This is likely unnecessary -- we only really need to give up if
+        # there are more then one *synthesized* union arguments.
```
[the remark this answers is not in the document]
Comment (Michael0x2a) [reply mypy/checkexpr.py]: Yeah, this was a can of worms I really didn't want to open.

I'll leave it up to somebody smarter then me to figure out how to handle cases like these if it ever becomes necessary.

File: mypy/subtypes.py [hunk 13]
```diff
@@ -568,16 +571,33 @@ def non_method_protocol_members(tp: TypeInfo) -> List[str]:
...
+    is_compat_return:
+        The check we want to run against the return type.
+        If None, use the 'is_compat' check.
+
+    check_args_covariantly:
+        If true, check if the left's args is compatible with the right's
+        instead of the other way around (contravariantly).
+
+        This function is mostly used to check if the left is a subtype of the right which
+        is why the default is to check the args contravariantly. However, it's occasionally
```
[the remark this answers is not in the document]
Comment (Michael0x2a) [reply mypy/subtypes.py]: Done

File: mypy/types.py [hunk 14]
```diff
@@ -674,6 +675,8 @@ class CallableType(FunctionLike):
                                  # 'dict')
                  'from_type_type',  # Was this callable generated by analyzing Type[...]
                                     # instantiation?
+                 'from_overloads',  # Was this callable generated by synthesizing
```
[the remark this answers is not in the document]
Comment (Michael0x2a) [reply mypy/types.py]: Whoops, forgot to delete that

File: test-data/unit/check-overloading.test [hunk 15]
```diff
@@ -1391,7 +1409,7 @@ def r(x: Any) -> Any:...
 @overload
 def g(x: A) -> A: ...
 @overload
-def g(x: Tuple[A1, ...]) -> B: ...  # E: Overloaded function signatures 2 and 3 overlap with incompatible return types
+def g(x: Tuple[A1, ...]) -> B: ...  # TODO: This should be an error
```
[the remark this answers is not in the document]
Comment (Michael0x2a) [reply test-data/unit/check-overloading.test]: I think the way to fix this in the general case is to add support for identifying partially overlapping types.

As an interim measure though, I added back in some of the existing code for checking overlaps -- it'll check for cases like these at least on the top level.

File: test-data/unit/check-overloading.test [hunk 16]
```diff
@@ -1504,3 +1522,428 @@ def bar(x: None) -> object: ...  # E: Overloaded function signatures 1 and 2 ove
...
+f2: None
+f3: Optional[Callable[[int], str]]
+
+reveal_type(mymap(f1, seq))  # E: Revealed type is 'typing.Iterable[builtins.str*]'
+reveal_type(mymap(f2, seq))  # E: Revealed type is 'typing.Iterable[builtins.int*]'
+reveal_type(mymap(f3, seq))  # E: Revealed type is 'Union[typing.Iterable[builtins.int], typing.Iterable[builtins.str]]'
+
+[builtins fixtures/list.pyi]
+[typing fixtures/typing-full.pyi]
+
```
[the remark this answers is not in the document]
Comment (Michael0x2a) [reply test-data/unit/check-overloading.test]: Done (I hope)

**Comment by Michael0x2a** [comment]
@ilevkivskyi -- I think this is ready for a second round of review whenever you have time.

Also, I'm not sure if this will actually make reviewing any easier, but I tried to split up my changes into self-contained commits w/ comments so you can check each distinct change + corresponding tests one-by-one instead of all at once if you prefer.

The biggest change I made was to redo how union-math handles generics. My previous implementation tried to use the caller arguments to infer types before unioning; the new logic tries to defer generics handling as much as possible and just unions together the typevars as they are. 

That said, my approach makes the assumption that typevars are internally "consistent" between overload alternatives: if a user uses a typevar T in one alternative and a typevar S in another in the same way, it treats those two TypeVars as distinct even though they could be unified. The `testOverloadInferUnionReturnWithInconsistentTypevarNames` test case has an example of this.

Let me know if you think this isn't a correct approach.

I also just went ahead and added code to special-case descriptors: I think it ended up being only like 2 lines of code, so it seemed like an easy win.

File: mypy/checker.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ilevkivskyi) [reply mypy/checker.py]: OK, let's keep assert for a moment, I will run `testpr` on our internal codebases.

File: test-data/unit/check-overloading.test [hunk 17]
```diff
@@ -1504,3 +1559,919 @@ def bar(x: None) -> object: ...  # E: Overloaded function signatures 1 and 2 ove
...
+    return x
+
+def test(x: T) -> T:
+    y: Union[str, int]
+
+    reveal_type(consistent(x, y))  # E: Revealed type is 'T`-1'
+
+    # TODO: Should we try and handle this differently?
+    # On one hand, this overload is defined in a weird way so it's arguably
+    # the user's fault; on the other, there's nothing overtly wrong with it.
```
Comment (ilevkivskyi) [inline test-data/unit/check-overloading.test]: Just open an issue about this so we don't forget about this corner case. Currently this looks too weird to care.

File: mypy/checkexpr.py [hunk 18]
```diff
@@ -2993,3 +3203,44 @@ def map_formals_to_actuals(caller_kinds: List[int],
         for actual in actuals:
             actual_to_formal[actual].append(formal)
     return actual_to_formal
+
+
+def merge_typevars_in_callables_by_name(
```
Comment (ilevkivskyi) [inline mypy/checkexpr.py]: Maybe we can avoid adding this function here, and instead ensure that during semantic analysis there is a single binding for a type variable across overload variants. This would be a clearer solution IMO, but the current code around `TypeVarScope` is a bit "opaque" and requires a refactoring anyway, so maybe it makes sense to this in a separate PR, up to you.

File: mypy/types.py [hunk 19]
```diff
@@ -734,8 +735,10 @@ def copy_modified(self,
                       line: int = _dummy,
                       column: int = _dummy,
                       is_ellipsis_args: bool = _dummy,
+                      implicit: bool = _dummy,
                       special_sig: Optional[str] = _dummy,
                       from_type_type: bool = _dummy,
+                      from_overloads: bool = _dummy,
```
Comment (ilevkivskyi) [inline mypy/types.py]: You forgot to remove this one.

File: test-data/unit/check-overloading.test [hunk 20]
```diff
@@ -1026,18 +1039,60 @@ g(1, 'foo')
...
+c: List[Any]
+d: Union[List[int], List[str]]
+e: List[bool]
+f: List[object]
+g: List[Union[int, str]]
+
+reveal_type(foo(a))
+reveal_type(foo(b))
+reveal_type(foo(c))
+reveal_type(foo(d))
```
Comment (ilevkivskyi) [inline test-data/unit/check-overloading.test]: Do we have a test that `[]` matches the first overload (as IIRC it is defined in your spec)? Also we probably need similar tests for empty user defined covariant collections.

File: test-data/unit/check-overloading.test [hunk 21]
```diff
@@ -1504,3 +1559,919 @@ def bar(x: None) -> object: ...  # E: Overloaded function signatures 1 and 2 ove
...
+def f(*, p2: B, p1: X) -> Y: ...  # E: Overloaded function signature 2 will never be matched: signature 1's parameter type(s) are the same or broader
+def f(*, p1, p2): ...
+
+@overload
+def g(*, p1: X, p2: B) -> X: ...  # E: Overloaded function signatures 1 and 2 overlap with incompatible return types
+@overload
+def g(*, p2: A, p1: X) -> Y: ...
+def g(*, p1, p2): ...
+
+[case testOverloadWithVariableArgsAreOverlapping-skip]
```
Comment (ilevkivskyi) [inline test-data/unit/check-overloading.test]: Some tests cases are too long, please consider splitting at least the few longest in parts.

File: test-data/unit/check-overloading.test [hunk 22]
```diff
@@ -1504,3 +1559,919 @@ def bar(x: None) -> object: ...  # E: Overloaded function signatures 1 and 2 ove
...
+    Dummy[T1]().foo(x3, y3)  # E: Argument 1 to "foo" of "Dummy" has incompatible type "Union[List[S], List[Tuple[S, T1]]]"; expected "Union[List[Tuple[T1, S]], List[S]]"
+
+    x4: Union[List[int], List[Tuple[C, int]]]
+    y4: int
+    reveal_type(Dummy[C]().foo(x4, y4))  # E: Revealed type is 'Union[__main__.C, builtins.int*]'
+    Dummy[A]().foo(x4, y4)               # E: Cannot infer type argument 1 of "foo" of "Dummy"
+
+    return arg1, arg2
+
+
```
Comment (ilevkivskyi) [inline test-data/unit/check-overloading.test]: Avoid double empty lines in tests.

**Comment by ilevkivskyi** [comment]
(Also note there is a minor merge conflict now.)
